File: sudoku.py
import random
import time
import os
import copy
from collections.abc import Callable
from typing import List, Tuple, Optional


class Sudoku:
    def __init__(self, size=9, solver: Callable = None):
        self.size = size
        self.box_size = int(size ** 0.5)
        self.board = self._create_empty_board()
        self.solver = solver
        self.steps = 0
        self.solution = None
        self.generate_puzzle()
    
    def _create_empty_board(self) -> List[List[int]]:
        return [[0 for _ in range(self.size)] for _ in range(self.size)]
    
    def generate_puzzle(self):
        self._fill_diagonal_boxes()
        self._solve_board(copy.deepcopy(self.board))
        solved_board = copy.deepcopy(self.board)
        self._remove_digits(solved_board)
    
    def _fill_diagonal_boxes(self):
        for i in range(0, self.size, self.box_size):
            self._fill_box(i, i)
    
    def _fill_box(self, row: int, col: int):
        numbers = list(range(1, self.size + 1))
        random.shuffle(numbers)
        
        index = 0
        for i in range(self.box_size):
            for j in range(self.box_size):
                self.board[row + i][col + j] = numbers[index]
                index += 1
    
    def _solve_board(self, board: List[List[int]]) -> bool:
        empty_cell = self._find_empty_cell(board)
        if not empty_cell:
            return True
        
        row, col = empty_cell
        for num in range(1, self.size + 1):
            if self._is_valid_placement(board, row, col, num):
                board[row][col] = num
                
                if self._solve_board(board):
                    self.board = board
                    return True
                
                board[row][col] = 0
        
        return False
    
    def _remove_digits(self, solved_board: List[List[int]], difficulty: float = 0.7):
        self.solution = copy.deepcopy(solved_board)
        cells_to_remove = int(self.size * self.size * difficulty)
        
        all_cells = [(i, j) for i in range(self.size) for j in range(self.size)]
        random.shuffle(all_cells)
        
        for i in range(min(cells_to_remove, len(all_cells))):
            row, col = all_cells[i]
            self.board[row][col] = 0
    
    def _find_empty_cell(self, board: List[List[int]]) -> Optional[Tuple[int, int]]:
        for i in range(self.size):
            for j in range(self.size):
                if board[i][j] == 0:
                    return (i, j)
        return None
    
    def _is_valid_placement(self, board: List[List[int]], row: int, col: int, num: int) -> bool:
        return (self._is_valid_in_row(board, row, num) and
                self._is_valid_in_column(board, col, num) and
                self._is_valid_in_box(board, row, col, num))
    
    def _is_valid_in_row(self, board: List[List[int]], row: int, num: int) -> bool:
        return num not in board[row]
    
    def _is_valid_in_column(self, board: List[List[int]], col: int, num: int) -> bool:
        return num not in [board[i][col] for i in range(self.size)]
    
    def _is_valid_in_box(self, board: List[List[int]], row: int, col: int, num: int) -> bool:
        box_row = row - row % self.box_size
        box_col = col - col % self.box_size
        
        for i in range(self.box_size):
            for j in range(self.box_size):
                if board[box_row + i][box_col + j] == num:
                    return False
        return True
    
    def display_board(self):
        os.system('cls' if os.name == 'nt' else 'clear')
        print(f"\n  SUDOKU SOLVER v1.0  |  Steps: {self.steps}\n")
        
        horizontal_line = self._create_horizontal_line()
        
        for i in range(self.size):
            if i % self.box_size == 0:
                print(horizontal_line)
            
            row_string = self._create_row_string(i)
            print(row_string)
        
        print(horizontal_line + "\n")
    
    def _create_horizontal_line(self) -> str:
        return "  " + "+".join(["-" * (self.box_size * 2 + 1)] * self.box_size)
    
    def _create_row_string(self, row_index: int) -> str:
        row_string = "  "
        
        for col_index in range(self.size):
            if col_index % self.box_size == 0:
                row_string += "| "
            
            cell_value = self.board[row_index][col_index]
            if cell_value == 0:
                row_string += ". "
            else:
                row_string += f"\033[97m{cell_value}\033[0m "
        
        return row_string + "|"
    
    def solve(self):
        if not self.solver:
            raise ValueError("No solver function provided")
        
        self.display_board()
        print("\nInitial board. Starting solver in 2 seconds...\n")
        time.sleep(2)
        
        return self.solver(self, self.display_board)


def brute_force_solver(sudoku: Sudoku, display_function):
    VISUALIZATION_DELAY = 0.1
    BACKTRACK_DELAY = 0.05
    
    def backtrack(row=0, col=0):
        sudoku.steps += 1
        
        if col == sudoku.size:
            return backtrack(row + 1, 0)
        
        if row == sudoku.size:
            return True
        
        if sudoku.board[row][col] != 0:
            return backtrack(row, col + 1)
        
        for num in range(1, sudoku.size + 1):
            if sudoku._is_valid_placement(sudoku.board, row, col, num):
                sudoku.board[row][col] = num
                
                display_function()
                time.sleep(VISUALIZATION_DELAY)
                
                if backtrack(row, col + 1):
                    return True
                
                sudoku.board[row][col] = 0
                display_function()
                time.sleep(BACKTRACK_DELAY)
        
        return False
    
    is_solved = backtrack()
    
    if is_solved:
        display_function()
        print(f"\nSolved successfully in {sudoku.steps} steps!")
    else:
        print("\nNo solution exists.")
    
    return is_solved


def main():
    print("\nINITIALIZING SUDOKU SOLVER...\n")
    time.sleep(1)
    
    sudoku = Sudoku(solver=brute_force_solver)
    sudoku.solve()


if __name__ == "__main__":
    main()

File: sudoku.cpp
#include <iostream>
#include <vector>
#include <random>
#include <algorithm>
#include <chrono>
#include <thread>
#include <functional>
#include <cstdlib>
#include <optional>
#include <numeric>

using namespace std;

class Sudoku {
private:
    static constexpr int DEFAULT_SIZE = 9;
    static constexpr double DEFAULT_DIFFICULTY = 0.7;
    static constexpr int VISUALIZATION_DELAY_MS = 100;
    static constexpr int BACKTRACK_DELAY_MS = 50;
    
    int size;
    int boxSize;
    vector<vector<int>> board;
    vector<vector<int>> solution;
    function<bool(Sudoku&, function<void()>)> solver;
    int steps;
    mt19937 rng;

public:
    explicit Sudoku(int boardSize = DEFAULT_SIZE, 
                   function<bool(Sudoku&, function<void()>)> solverFunc = nullptr)
        : size(boardSize), boxSize(static_cast<int>(sqrt(boardSize))), 
          board(createEmptyBoard()), solver(solverFunc), steps(0),
          rng(chrono::steady_clock::now().time_since_epoch().count()) {
        generatePuzzle();
    }

    bool solve() {
        if (!solver) {
            throw runtime_error("No solver function provided");
        }
        
        displayBoard();
        cout << "\nInitial board. Starting solver in 2 seconds...\n" << endl;
        this_thread::sleep_for(chrono::seconds(2));
        
        return solver(*this, [this]() { displayBoard(); });
    }

    void displayBoard() const {
        system("clear");
        cout << "\n  SUDOKU SOLVER v1.0  |  Steps: " << steps << "\n" << endl;
        
        string horizontalLine = createHorizontalLine();
        
        for (int i = 0; i < size; ++i) {
            if (i % boxSize == 0) {
                cout << horizontalLine << endl;
            }
            
            cout << createRowString(i) << endl;
        }
        
        cout << horizontalLine << "\n" << endl;
    }

    vector<vector<int>>& getBoard() { return board; }
    const vector<vector<int>>& getBoard() const { return board; }
    int getSize() const { return size; }
    int& getSteps() { return steps; }
    bool isValidPlacement(int row, int col, int num) const {
        return isValidInRow(row, num) && 
               isValidInColumn(col, num) && 
               isValidInBox(row, col, num);
    }

private:
    vector<vector<int>> createEmptyBoard() const {
        return vector<vector<int>>(size, vector<int>(size, 0));
    }

    void generatePuzzle() {
        fillDiagonalBoxes();
        solveBoard(board);
        solution = board;
        removeDigits(DEFAULT_DIFFICULTY);
    }

    void fillDiagonalBoxes() {
        for (int i = 0; i < size; i += boxSize) {
            fillBox(i, i);
        }
    }

    void fillBox(int startRow, int startCol) {
        vector<int> numbers(size);
        iota(numbers.begin(), numbers.end(), 1);
        shuffle(numbers.begin(), numbers.end(), rng);

        int index = 0;
        for (int i = 0; i < boxSize; ++i) {
            for (int j = 0; j < boxSize; ++j) {
                board[startRow + i][startCol + j] = numbers[index++];
            }
        }
    }

    bool solveBoard(vector<vector<int>>& targetBoard) {
        auto emptyCell = findEmptyCell(targetBoard);
        if (!emptyCell.has_value()) {
            return true;
        }

        auto [row, col] = emptyCell.value();
        for (int num = 1; num <= size; ++num) {
            if (isValidPlacement(targetBoard, row, col, num)) {
                targetBoard[row][col] = num;
                
                if (solveBoard(targetBoard)) {
                    board = targetBoard;
                    return true;
                }
                
                targetBoard[row][col] = 0;
            }
        }
        
        return false;
    }

    void removeDigits(double difficulty) {
        int cellsToRemove = static_cast<int>(size * size * difficulty);
        
        vector<pair<int, int>> allCells;
        for (int i = 0; i < size; ++i) {
            for (int j = 0; j < size; ++j) {
                allCells.emplace_back(i, j);
            }
        }
        
        shuffle(allCells.begin(), allCells.end(), rng);
        
        for (int i = 0; i < min(cellsToRemove, static_cast<int>(allCells.size())); ++i) {
            auto [row, col] = allCells[i];
            board[row][col] = 0;
        }
    }

    optional<pair<int, int>> findEmptyCell(const vector<vector<int>>& targetBoard) const {
        for (int i = 0; i < size; ++i) {
            for (int j = 0; j < size; ++j) {
                if (targetBoard[i][j] == 0) {
                    return make_pair(i, j);
                }
            }
        }
        return nullopt;
    }

    bool isValidPlacement(const vector<vector<int>>& targetBoard, int row, int col, int num) const {
        return isValidInRow(targetBoard, row, num) && 
               isValidInColumn(targetBoard, col, num) && 
               isValidInBox(targetBoard, row, col, num);
    }

    bool isValidInRow(const vector<vector<int>>& targetBoard, int row, int num) const {
        return find(targetBoard[row].begin(), targetBoard[row].end(), num) == targetBoard[row].end();
    }

    bool isValidInColumn(const vector<vector<int>>& targetBoard, int col, int num) const {
        for (int i = 0; i < size; ++i) {
            if (targetBoard[i][col] == num) {
                return false;
            }
        }
        return true;
    }

    bool isValidInBox(const vector<vector<int>>& targetBoard, int row, int col, int num) const {
        int boxRow = row - row % boxSize;
        int boxCol = col - col % boxSize;
        
        for (int i = 0; i < boxSize; ++i) {
            for (int j = 0; j < boxSize; ++j) {
                if (targetBoard[boxRow + i][boxCol + j] == num) {
                    return false;
                }
            }
        }
        return true;
    }

    bool isValidInRow(int row, int num) const {
        return isValidInRow(board, row, num);
    }

    bool isValidInColumn(int col, int num) const {
        return isValidInColumn(board, col, num);
    }

    bool isValidInBox(int row, int col, int num) const {
        return isValidInBox(board, row, col, num);
    }

    string createHorizontalLine() const {
        string line = "  ";
        for (int i = 0; i < boxSize; ++i) {
            if (i > 0) line += "+";
            line += string(boxSize * 2 + 1, '-');
        }
        return line;
    }

    string createRowString(int rowIndex) const {
        string rowStr = "  ";
        
        for (int colIndex = 0; colIndex < size; ++colIndex) {
            if (colIndex % boxSize == 0) {
                rowStr += "| ";
            }
            
            int cellValue = board[rowIndex][colIndex];
            if (cellValue == 0) {
                rowStr += ". ";
            } else {
                rowStr += "\033[97m" + to_string(cellValue) + "\033[0m ";
            }
        }
        
        return rowStr + "|";
    }
};

bool bruteForceSolver(Sudoku& sudoku, function<void()> displayFunction) {
    function<bool(int, int)> backtrack = [&](int row, int col) -> bool {
        sudoku.getSteps()++;
        
        if (col == sudoku.getSize()) {
            return backtrack(row + 1, 0);
        }
        
        if (row == sudoku.getSize()) {
            return true;
        }
        
        if (sudoku.getBoard()[row][col] != 0) {
            return backtrack(row, col + 1);
        }
        
        for (int num = 1; num <= sudoku.getSize(); ++num) {
            if (sudoku.isValidPlacement(row, col, num)) {
                sudoku.getBoard()[row][col] = num;
                
                displayFunction();
                this_thread::sleep_for(chrono::milliseconds(100));
                
                if (backtrack(row, col + 1)) {
                    return true;
                }
                
                sudoku.getBoard()[row][col] = 0;
                displayFunction();
                this_thread::sleep_for(chrono::milliseconds(50));
            }
        }
        
        return false;
    };
    
    bool isSolved = backtrack(0, 0);
    
    if (isSolved) {
        displayFunction();
        cout << "\nSolved successfully in " << sudoku.getSteps() << " steps!" << endl;
    } else {
        cout << "\nNo solution exists." << endl;
    }
    
    return isSolved;
}

int main() {
    cout << "\nINITIALIZING SUDOKU SOLVER...\n" << endl;
    this_thread::sleep_for(chrono::seconds(1));
    
    Sudoku sudoku(9, bruteForceSolver);
    sudoku.solve();
    
    return 0;
}